src/gui/Lv2ViewBase.cpp

```diff
@@ -0,0 +1,251 @@
+/*
+ * Lv2ViewBase.cpp - base class for Lv2 plugin views
+ *
+ * Copyright (c) 2018-2019 Johannes Lorenz <j.git$$$lorenz-ho.me, $$$=@>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "Lv2ViewBase.h"
+
+#ifdef LMMS_HAVE_LV2
+
+#include <QGridLayout>
+#include <QGroupBox>
+#include <QMdiSubWindow>
+#include <QPushButton>
+#include <QHBoxLayout>
+#include <lilv/lilv.h>
+
+#include "Controls.h"
+#include "Engine.h"
+#include "GuiApplication.h"
+#include "embed.h"
+#include "gui_templates.h"
+#include "LedCheckbox.h"
+#include "Lv2ControlBase.h"
+#include "Lv2Manager.h"
+#include "Lv2Proc.h"
+#include "Lv2Ports.h"
+#include "MainWindow.h"
+#include "SubWindow.h"
+
+
+
+
+Lv2ViewProc::Lv2ViewProc(QWidget* parent, Lv2Proc* ctrlBase,
+	int colNum, int curProc, int nProc, const QString& name) :
+	LinkedModelGroupViewBase (parent, ctrlBase->isLinking(), colNum,
+								curProc, nProc, name)
+{
+	class SetupWidget : public Lv2Ports::Visitor
+	{
+		AutoLilvNode commentUri = uri(LILV_NS_RDFS "comment");
+	public:
+		QWidget* par; // input
+		ControlBase* control = nullptr; // output
+		void visit(Lv2Ports::Control& port) override
+		{
+			if(port.m_flow == Lv2Ports::Flow::Input)
+			{
+				using PortVis = Lv2Ports::Vis;
+
+				switch (port.m_vis)
+				{
+					case PortVis::None:
+						control = new KnobControl(par);
+						break;
+					case PortVis::Integer:
+						control = new LcdControl((port.m_max <= 9.0f) ? 1 : 2,
+													par);
+						break;
+					case PortVis::Enumeration:
+						control = new ComboControl(par);
+						break;
+					case PortVis::Toggled:
+						control = new CheckControl(par);
+						break;
+				}
+				control->setText(port.name());
+
+				LilvNodes* props = lilv_port_get_value(
+					port.m_plugin, port.m_port, commentUri.get());
+				LILV_FOREACH(nodes, itr, props)
```

Comment:
space?

Reply:
It's a macro call, i.e. like a function call. As we've just agreed on discord, we don't change it.